### DISCUSSION TITLE
Embedding MOOSE into a C++ library

### DISCUSSION THREAD

**Discussion by felixcrazzolara** [idaholab/moose#27465] [Q&A General] [answered]
First of all, thank you for providing and maintaining this wonderful piece of software!

I would like to do reinforcement learning with MOOSE. Ideally, I'd like to have a Jupyter notebook to easily experiment with it. The Jupyter notebook should import some Python modules which in turn uses C++ code, either connecting the dots with Pybind11 or CPython.

Is it easily feasible to embed MOOSE into a C++ library? I want to run a tensor_mechanics simulation. The RL agent decides in each timestep what forces to apply to the structure. Finally, I'd like to get all data immediately in memory. Nothing should be saved to CSV files or so.

Alternatively, if MOOSE is really meant to be a standalone application, I could maybe inject some code that communicates with a Python server/client using UNIX/TCP sockets to pass data back and forth.

What are your thoughts on this? What is the most natural approach to go for given the code structure of MOOSE?

**Comment by GiudGiud** [comment, marked as answer]
Hello

We have an optimization and a stochastic tools (linked with libtorch) module to perform these studies inside the framework, rather than calling MOOSE from Python. We do that internally because of performance concerns/ 

>  I could maybe inject some code that communicates with a Python server/client using UNIX/TCP sockets to pass data back and forth.

you are welcome to do that. You can also work with the MooseServer class to perform these more dynamic communications

Guillaume

**Reply by felixcrazzolara** [reply]
Thank you that sounds great! Also, for people trying to compile Moose with libtorch support, it worked for me by building it from source, but using the same conda environment as the one of Moose. Tried the everything in the official documentation of Moose on libtorch and some other stuff, but nothing of that worked out for me.

**Reply by ze897c** [reply]
Hi folks,

What is the actual answer to “Is it easily feasible to embed MOOSE into a C++ library?” I have need to do something similar and for various reasons we are not inclined to develop _within_ the MOOSE framework. If the suggested path is to use the MooseServer class, is there a good example to follow?

Essentially what I would like to be able to do is to use MOOSE as a solver on a persistent mesh with material properties that are varied by the parent C++ library. I would like to keep MOOSE “running” and let the parent non-MOOSE app control the flow. I hope this makes some sense. All suggestions are welcome. Thanks!
Matt

**Reply by GiudGiud** [reply]
hello

It is feasible. I think the main difficulty will be working out the makefiles (or the dynamic loading). Is your project using make or cmake?

MooseServer class will only go so far but it can be extended. Feel free to start with this solution for now if you don't want to work on the build system.

> I would like to keep MOOSE “running” and let the parent non-MOOSE app control the flow. I hope this makes some sense.

This makes sense. This is not typically what we do when the other application is only computing properties. We typically have MOOSE run the show because of the time step decisions, the coupling to other apps, etc.
But the webservercontrol is a product of a project choosing to run the simulation from Python.

>  is there a good example to follow?

There is a test here
https://github.com/idaholab/moose/tree/next/test/tests/controls/web_server_control
@loganharbour is there an example with the webserver control from the depletion project? I did not find one in Cardinal

**Reply by loganharbour** [reply]
These are the best two references:

- https://mooseframework.inl.gov/source/controls/WebServerControl.html
- https://mooseframework.inl.gov/python/MooseControl/index.html

**Reply by ze897c** [reply]
Hi folks.

Thank you VERY much for the prompt reply. I will start here.

Cheers!!
Matt
